packages/tpetra/core/src/Tpetra_CrsSingletonFilter_LinearProblem_def.hpp

```diff
@@ -0,0 +1,1768 @@
+  Kokkos::View<size_t*, device_type> numView;
+
+  // constructor for counting
+  GenerateReducedMapFunctor(int color_, Map_color_type mapColors_)
+    : color(color_)
+    , mapColors(mapColors_) {}
+
+  // constructor for inserting
+  GenerateReducedMapFunctor(int color_, Map_color_type mapColors_, LocalMap_type lclMap_, IndexList_type indexList_)
+    : color(color_)
+    , mapColors(mapColors_)
+    , lclMap(lclMap_)
+    , indexList(indexList_) {
+    Kokkos::resize(numView, 1);
+    Kokkos::deep_copy(numView, 0);
+  }
+
+  // Functor to count number of elements with matching color (parallel-reduce)
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const size_t i, size_t& num) const {
+    if (mapColors(i, 0) == color) {
+      num++;
+    }
+  }
+
+  // Functor to insert elements with matching color (parallel-for)
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const size_t i) const {
+    if (mapColors(i, 0) == color) {
+      auto num       = Kokkos::atomic_fetch_add(&numView(0), 1);  // ..atomic to count..
+      indexList(num) = lclMap.getGlobalElement(i);                // to global
+    }
+  }
+};
+
+//==============================================================================
+
+template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+CrsSingletonFilter_LinearProblem<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+    CrsSingletonFilter_LinearProblem(bool run_on_host, bool verbose)
+  : globalNumSingletonRows_(Teuchos::as<local_ordinal_type>(0))
+  , globalNumSingletonCols_(Teuchos::as<local_ordinal_type>(0))
+  , RatioOfDimensions_(0.0)
+  , RatioOfNonzeros_(0.0)
+  , HaveReducedProblem_(false)
+  , AnalysisDone_(false)
+  , SymmetricElimination_(true)
+  , localMaxNumRowEntries_(Teuchos::as<local_ordinal_type>(0))
+  , FullMatrixIsCrsMatrix_(false)
+  , run_on_host_(run_on_host)
+  , verbose_(verbose) {
+}
+
+//==============================================================================
+
+template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+typename CrsSingletonFilter_LinearProblem<Scalar, LocalOrdinal, GlobalOrdinal, Node>::NewType
+CrsSingletonFilter_LinearProblem<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+operator()(const CrsSingletonFilter_LinearProblem::OriginalType& originalLinearProblem) {
+  analyze(originalLinearProblem);
+  return construct();
+}
+
+//==============================================================================
+
+template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void CrsSingletonFilter_LinearProblem<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+    analyze(const CrsSingletonFilter_LinearProblem::OriginalType& originalLinearProblem) {
+  this->origObj_ = originalLinearProblem;
+
+  FullMatrix_ = originalLinearProblem->getMatrix();
+
+  Analyze(FullMatrix());
+
+  if (verbose_ && FullMatrix()->getComm()->getRank() == 0) {
+    std::cout << "\nAnalyzed Singleton Problem:\n";
```

Comment:
Maybe in the future: use a member variable for the output stream instead of cout.